sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1162 @@
+<<queue-state, state>> is introduced to queue objects, allowing an instance to be
+put into a mode where command-groups are recorded to a graph rather than
+submitted immediately for execution.
+
+<<new-queue-member-functions, Three new member functions>> are also added to the
+`sycl::queue` class with this extension. Two functions for selecting the state
+of the queue, and another function for submitting a graph to the queue.
+
+==== Queue State
+
+:queue-info-table: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.queue.info
+
+The `sycl::queue` object can be in either of two states. The default
+`queue_state::executing` state is where the queue has its normal semantics of
+submitted command-groups being immediately scheduled for asynchronous execution.
+
+The alternative `queue_state::recording` state is used for graph construction.
+Instead of being scheduled for execution, command-groups submitted to the queue
+are recorded to a graph object as new nodes for each submission. After recording
+has finished and the queue returns to the executing state, the recorded commands are
+not then executed, they are transparent to any following queue operations.
+
+.Queue State
+image::images/sycl_ext_oneapi_graph/queue-state.svg[]
+
+The state of a queue can be queried with `queue::get_info` using template
+parameter `info::queue::state`. The following entry is added to the
+{queue-info-table}[queue info table] to define this query:
+
+Table 9. Queue info query
+[cols="2a,a,a"]
+|===
+| Queue Descriptors | Return Type | Description
+
+| `info::queue::state`
+| `ext::oneapi::experimental::queue_state`
+| Returns the state of the queue
+
+|===
+
+A default constructed event is returned when a user submits a command-group to
+a queue in the recording state. These events have status
+`info::event_command_status::complete` and a user waiting on them will return
+immediately.
+
+==== Queue Properties
+
+:queue-properties: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:queue-properties
+
+There are {queue-properties}[two properties] defined by the core SYCL
+specification that can be passed to a `sycl::queue` on construction via the
+property list parameter. They interact with this extension in the following
+ways:
+
+1. `property::queue::in_order` - When a queue is created with the in-order
+   property, recording its operations results in a straight-line graph, as each
+   operation has an implicit dependency on the previous operation. However,
+   a graph submitted to an in-order queue will keep its existing structure such
+   that the complete graph executes in-order with respect to the other
+   command-groups submitted to the queue.
+
+2. `property::queue::enable_profiling` - This property has no effect on graph
+   recording. When set on the queue a graph is submitted to however, it allows
+   profiling information to be obtained from the event returned by a graph
+   submission.
```

Comment:
What exactly does this profiling information represent?  The SYCL 2020 functionality is all defined in terms of the "command group" in this table: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.event.profilinginfo

Since a graph may execute multiple kernels, copies, etc, how should a developer reason about these?  Is `command_submit` the time that the graph is submitted, `command_start` the time that the first node begins executing, and `command_end` the time that the last leaf node ends executing?

Might `command_end` - `command_start` give a pessimistic view of execution time, since the device may decide to schedule other work (i.e. outside the graph) between scheduling nodes from the same graph?

Reply:
Created a PR to clarify this https://github.com/reble/llvm/pull/88, but your suggested semantics for each of the queries is what I've specified. Although it may give a pessimistic view of execution time, I think it'll be hard for us to guarantee better across all cases. If a graph's characteristics allow it to be mapped to a single PI command-list then that's great, but more complicated graphs may need partitioning into many underlying PI command-lists, which makes it hard to specify that we'll profile the graph as an atomic unit.

We do have the idea of adding some sort of profiling node in the future for more fine-grained profiling by the user, but that really is just an idea at this point.